evaluate_agent.py

```diff
@@ -0,0 +1,125 @@
+@click.option("--agentdirectory", "-a", default='./agent', help="Path to the directory containing the agent that should be evaluated.", type=click.Path(exists=True))
+@click.option("--save/--dont-save", default=False, help="Boolean indicating whether the intermediate results of the evaluation should be saved.")
+def evaluate(gpu, imagefile, boxfile, agentdirectory, save):
+    max_steps_per_image = 200
+    max_steps_per_run = 40
+    max_trigger_events_per_image = 1
+
+    relative_paths = np.loadtxt(imagefile, dtype=str)
+    images_base_path = os.path.dirname(imagefile)
+    absolute_paths = [images_base_path + i.strip('.') for i in relative_paths]
+
+    images = np.loadtxt(imagefile, dtype=str).tolist()
+
+    gt_bboxes = np.load(boxfile)
+
+    env = TextLocEnv(absolute_paths, gt_bboxes, gpu)
+    agent = load_agent(env, agentdirectory, gpu)
+
+    pred_bboxes = []
+
+    pbar = tqdm(images, unit="images")
+    for image_index, image in enumerate(pbar):
+        predicted_bboxes_for_image = []
+        _ = env.reset(image_index)
+        steps_in_current_image = 1
+        trigger_events_in_current_image = 0
+        while steps_in_current_image <= max_steps_per_image \
+                and trigger_events_in_current_image < max_trigger_events_per_image:
+            obs = env.reset(image_index, True)
+            steps_in_current_run = 1
+            while steps_in_current_run <= max_steps_per_run:
+                pbar.set_postfix(image=image[-11:], step=steps_in_current_image)
+                action = agent.act(obs)
+                obs, r, done, _ = env.step(action)
+                if done:
+                    predicted_bboxes_for_image.append(env.bbox)
+                    trigger_events_in_current_image += 1
+                    break
+                steps_in_current_run += 1
+                steps_in_current_image += 1
+        pred_bboxes.append(predicted_bboxes_for_image)
+
+    gt_bboxes = list(map(convert_gt_bboxes_to_chainercv_format, gt_bboxes))
+    gt_labels = []
+    for bbox in gt_bboxes:
+        labels = []
+        for _ in range(len(bbox)):
+            labels.append(True)
+        gt_labels.append(labels)
+
+    pred_bboxes = list(map(convert_env_bboxes_to_chainercv_format, pred_bboxes))
+    pred_labels = []
+    pred_scores = []
+    for bbox in pred_bboxes:
+        labels = []
+        scores = []
+        for _ in range(len(bbox)):
+            labels.append(True)
+            scores.append(1.0)
+        pred_labels.append(labels)
+        scores = np.array(scores, dtype=np.float32)
+        pred_scores.append(scores)
+
+    if save:
+        np.save('gt_bboxes.npy', gt_bboxes)
+        np.save('gt_labels.npy', gt_labels)
+        np.save('pred_bboxes.npy', pred_bboxes)
+        np.save('pred_labels.npy', pred_labels)
+        np.save('pred_scores.npy', pred_scores)
+
+    eval = eval_detection_voc(pred_bboxes, pred_labels, pred_scores, gt_bboxes, gt_labels)
```

Comment:
Ok, in our case this is not a problem because we have only one bounding box, but if there was more than one how does this precision metric deal with it? I mean Assume that you have 5 gt_bboxes in a given image and assume that the agent finds them all with IoU of 1.0, but it finds them in a different order (say, 2,4,3,0,1). Will this be counted as precision of 100% (correct precision) or will it think that it got them all wrong?

Reply:
Actually I'm not too sure about that either. According to the [documentation comments inside `eval_detection_voc`'s source code](https://github.com/chainer/chainercv/blob/master/chainercv/evaluations/eval_detection_voc.py#L11), I'd assume that it should work for multiple bounding boxes per image as well. I haven't tested this, though.

Reply:
As discussed: I’m marking this as resolved since you successfully tested it. Thanks!